python/arcticdb/diagnostics_engine/version_list_diagnosis.py

```diff
@@ -0,0 +1,342 @@
+        is_last_node_and_exists_compacted = compacted_versions and i == len(parsed_keys) - 1
+
+        version_exists = (
+            current_key_version in parsed_versions and parsed_versions[current_key_version]["deleted"] == False
+        )
+
+        # Get node color and label
+        node_color, node_label = get_node_properties(
+            current_key_version, seen_nodes, version_exists, is_last_node_and_exists_compacted
+        )
+
+        node_colors.append(node_color)
+        node_labels[node_name] = node_label
+
+        # We need to mark the seen nodes to establish when a node is a tombstone and when it denotes the creation of a new version.
+        seen_nodes[current_key_version] = True
+
+        # Break loop if steps are 0 or current key version is target
+        if steps == 0 or current_key_version == target:
+            break
+
+        # Add edge to next node
+        if i < len(parsed_keys) - 1:
+            next_node_name = f"{parsed_keys[i + 1].version_id}_{i + 1}"
+            G.add_edge(node_name, next_node_name)
+
+    plt.figure(figsize=(12, 8))
+    pos = zigzag_layout(G)
+    nx.draw(G, pos, labels=node_labels, node_size=1200, arrows=True, node_color=node_colors)
+    add_graph_description(compacted_versions, exists_in_version_list, is_tombstoned)
+
+
+def parse_versions(versions):
+    """
+    Parse versions into a dictionary for easy access.
+    """
+    parsed_versions = {v["version"]: v for v in versions}
+    return parsed_versions
+
+
+def parse_keys(keys):
+    """
+    Parse and sort keys based on creation timestamp.
+    """
+    parsed_keys = sorted(keys, key=lambda k: k.creation_ts, reverse=True)
+    return parsed_keys
+
+
+def get_target_version(parsed_versions, parsed_keys, as_of):
+    """
+    Get the target version based on the 'as_of', or the latest version if 'as_of' is None.
+    """
+    if as_of is not None:
+        return as_of
+
+    alive_versions = parsed_versions.keys()
+    for key in parsed_keys:
+        if key in alive_versions and parsed_versions[key.version_id]["deleted"] == False:
+            return key.version_id
+
+
+def traverse_version_list(parsed_keys, target):
+    """
+    Traverse the version list to find the target version and the number of steps required to reach it.
+    """
+    steps = 0
+    is_present = False
+    for i, key in enumerate(parsed_keys):
+        steps += 1
+        if key.version_id == target:
+            is_present = True
+            break
+
+    return steps - 1, is_present
+
+
+def get_compacted_versions(parsed_versions, last_version_in_list):
+    """
+    Get all the compacted versions after the last node from the version list.
+    """
+    compacted_versions = []
+    for version in parsed_versions.keys():
+        if parsed_versions[version]["deleted"] == False and version <= last_version_in_list:
+            compacted_versions.append(version)
+    return compacted_versions
```

Comment:
Is this right? Does version compaction always compact everything into one object, including previously compacted objects?

Reply:
Yes, I conducted some experiments myself and also confirmed with @willdealtry. All the compacted versions are included in one version, which is always the last one in the chain.

Reply:
Compaction always compacts everything at the moment, from the second key in the version list. It doesn't have to - the version list would work fine if it was large chunks of data connected by keys that act as a pointer to the next one like a std::deque rather than a std::vector, but that's not implemented in the compaction at the moment.

https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/version/version_map.hpp#L374

Reply:
Cool, I wasn't certain of the behaviour, just wanted to make sure it had been tested